File: frontend/src/services/savedListingService.ts
import apiClient from './apiClient';
import type { PageResponse } from '../types/api';
import type { ListingSummaryResponse } from '../types/listing';

export const savedListingService = {
  /**
   * GET /saved-listings
   * Get user's saved listings (paginated)
   */
  getSavedListings: async (params?: { page?: number; size?: number }): Promise<PageResponse<ListingSummaryResponse>> => {
    const res = await apiClient.get<PageResponse<ListingSummaryResponse>>('/saved-listings', { params });
    return res.data;
  },

  /**
   * POST /saved-listings/:listingId
   * Toggle save status for a listing
   */
  toggleSave: async (listingId: number): Promise<{ saved: boolean; message: string }> => {
    const res = await apiClient.post<{ saved: boolean; message: string }>(`/saved-listings/${listingId}`);
    return res.data;
  },

  /**
   * GET /saved-listings/:listingId/check
   * Check if a listing is saved by current user
   */
  checkSaved: async (listingId: number): Promise<{ saved: boolean }> => {
    const res = await apiClient.get<{ saved: boolean }>(`/saved-listings/${listingId}/check`);
    return res.data;
  }
};

export default savedListingService;

File: core-service/src/main/java/com/totoro/listing/service/SavedListingService.java
package com.totoro.listing.service;

import com.totoro.listing.dto.ListingSummaryResponse;
import com.totoro.listing.dto.PageResponse;
import com.totoro.listing.entity.Listing;
import com.totoro.listing.entity.SavedListing;
import com.totoro.listing.repository.ListingRepository;
import com.totoro.listing.repository.SavedListingRepository;
import com.totoro.common.dto.UserProfileDto;
import com.totoro.internal.user.UserServiceClient;
import com.totoro.common.dto.UserProfileDto;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SavedListingService {

    private final SavedListingRepository savedListingRepository;
    private final ListingRepository listingRepository;
    private final UserServiceClient userServiceClient;
    private final ListingService listingService;

    /**
     * Toggle save/unsave a listing for the current user.
     * Returns true if saved, false if unsaved.
     */
    @Transactional
    public boolean toggleSave(Long userId, Long listingId) {
        UserProfileDto user = userServiceClient.getUserProfile(userId);
        Listing listing = listingRepository.findById(listingId)
                .orElseThrow(() -> new IllegalArgumentException("Listing not found"));

        if (savedListingRepository.existsByUserIdAndListingId(user.getId(), listingId)) {
            savedListingRepository.deleteByUserIdAndListingId(user.getId(), listingId);
            return false; // unsaved
        } else {
            SavedListing saved = SavedListing.builder()
                    .userId(user.getId())
                    .listing(listing)
                    .build();
            savedListingRepository.save(saved);
            return true; // saved
        }
    }

    public PageResponse<ListingSummaryResponse> getSavedListings(Long userId, Pageable pageable) {
        UserProfileDto user = userServiceClient.getUserProfile(userId);
        Page<SavedListing> page = savedListingRepository.findByUserId(user.getId(), pageable);

        List<ListingSummaryResponse> content = page.getContent().stream()
                .map(sl -> listingService.toSummaryResponse(sl.getListing()))
                .collect(Collectors.toList());

        return PageResponse.<ListingSummaryResponse>builder()
                .content(content)
                .page(page.getNumber())
                .size(page.getSize())
                .totalElements(page.getTotalElements())
                .totalPages(page.getTotalPages())
                .last(page.isLast())
                .build();
    }

    public boolean isListingSaved(Long userId, Long listingId) {
        UserProfileDto user = userServiceClient.getUserProfile(userId);
        return savedListingRepository.existsByUserIdAndListingId(user.getId(), listingId);
    }
}
